fhir-flow/src/main/java/com/ibm/fhir/flow/impl/UpstreamFHIRHistoryReader.java

```diff
@@ -0,0 +1,291 @@
+                    break;
+                }
+            }
+        }
+    }
+
+    /**
+     * Process the response bundle from the whole-system history call
+     * @param bundle
+     */
+    private void processBundle(Bundle bundle) {
+        if (bundle.getEntry().isEmpty()) {
+            return;
+        }
+
+        // Add all the tracker entries in one go - more efficient locking
+        List<ITrackerTicket> tickets = addTrackerValues(bundle);
+        int index = 0;
+        for (Bundle.Entry entry: bundle.getEntry()) {
+            ITrackerTicket trackerTicket = tickets.get(index++);
+            // The entry.id field is database resourceId which is used as the PK for the
+            // changes table. It just needs to be unique and match the change sequence. It
+            // does not need to be contiguous
+            long changeId = Long.parseLong(entry.getId());
+            switch (entry.getRequest().getMethod().getValueAsEnum()) {
+            case DELETE:
+                // no value to fetch, so just submit the interaction
+                flowWriter.submit(new Delete(changeId, trackerTicket, ResourceIdentifier.from(entry.getFullUrl())));
+                break;
+            case PUT:
+            case POST:
+                // initiate the request to read the resource. Because we need the version, we have to
+                // use the response.location field
+                Entry.Response response = entry.getResponse();
+                if (response != null) {
+                    ResourceIdentifierVersion riv = ResourceIdentifierVersion.from(response.getLocation());
+                    CompletableFuture<FlowFetchResult> flowFetchFuture = flowPool.requestResource(riv);
+    
+                    // and submit the interaction using the future which can be resolved later
+                    flowWriter.submit(new CreateOrUpdate(changeId, trackerTicket, riv, flowFetchFuture));
+                } else {
+                    throw new IllegalStateException("entry does not include a response: " + entry.toString());
+                }
+                break;
+            default:
+                break;
+            }
+        }
+    }
+
+    /**
+     * Add all the change id values from the bundle entries. We do this all at once
+     * because it means we only have to synchronize on the changeTrackerMap once and
+     * still the lock very short amount of time.
+     * @param bundle
+     */
+    private List<ITrackerTicket> addTrackerValues(Bundle bundle) {
+
+        List<Long> requestIds = new ArrayList<>(bundle.getEntry().size());
+        for (Bundle.Entry entry: bundle.getEntry()) {
+            // The entry.id field is database resourceId which is used as the PK for the
+            // changes table. It just needs to be unique and match the change sequence. It
+            // does not need to be contiguous
+            requestIds.add(Long.parseLong(entry.getId()));
```

Comment:
just wanted to flag this as another case where its dependent on our server's unique behavior.  probably fine, but thought it was worth mentioning.

Reply:
yeah, perhaps this is something we should bring up with the spec when dealing with Bundles and multiple pages of results. Right now, the id only needs to be unique within a Bundle...but for pagination I think there are additional uniqueness/sequence semantics which could be useful like above.

Reply:
I don't think HL7 will be interested in further constraining the paging behavior.
They base theirs on RFC 5005 and they keep it rather opaque...everyone does it a little different.

The ids we set there have no defined meaning int the spec...those fields are really for "internal linking"
https://www.hl7.org/fhir/element-definitions.html#Element.id but in our case its kind of handy to emit this extra info without needing to define an extension for it with some defined semantics.  IMHO clients really shouldn't depend on it unless we document it.